Add a simple token parser service

m-vo opened the pull request:
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes none (followup to #1509)
| Docs PR or issue | todo

**How it works** (updated):

This refactors the simple tokens parser into a service. It's called `contao.util.simple_token_parser` and lives in the `Util` namespace. It's accompanied by a custom `SimpleTokenExpressionLanguage`   that allows registering extension functions and is 'secure' (meaning it always overwrites the `constant()` function)*.

To customize the EL people can now create extensions by simply implementing the `ExpressionFunctionProviderInterface` and tagging it with `'contao.simple_token_extension`.

* I deprecated `StringUtil::parseSimpleTokens` and rewired all of it's usage to the new service.
* All existing tests are moved. I added an additional additional functional one that tests registering a custom EL extension.

*) We could now also create an extension ourselves for this part, but personally I like this more because this way it's always applied last. :man_shrugging: 


~~**CS Fixer Caveat**~~ Solution see 40a70c7b376d31d4dd98f7e175a7a706b69171f7

There is one part, I'm not sure how to solve (easily): the code contains two comparisons with type juggling that must stay this way, but the CS fixer tries to eliminate them on each run:

```php
switch ($operator) {
    case '==':
        return $tokenValue == $value;

     case '!=':
        return $tokenValue != $value;

     //...
```

m-vo committed:
move simple token parser into a service

m-vo committed:
move simple token parser into a service

m-vo committed:
move simple token parser tests

m-vo committed:
move simple token parser tests

m-vo committed:
add test for using custom EL function while parsing simple tokens

m-vo committed:
add test for using custom EL function while parsing simple tokens

m-vo committed:
rewire usage

m-vo committed:
rewire usage

Toflar commented on core-bundle/src/Resources/config/services.yml:
```diff
@@ -769,6 +769,12 @@ services:
             - [setName, [contao_frontend]]
 
+    contao.simple_tokens.parser:
+        class: Contao\CoreBundle\SimpleTokens\Parser
```
I would put it in the `Util` namespace and name the class `SimpleTokenParser`. And then that makes for `contao.util.simple_token_parser`.

Toflar commented on core-bundle/src/SimpleTokens/Parser.php:
```diff
@@ -0,0 +1,231 @@
+
+    public function __construct(LoggerInterface $logger = null, ExpressionLanguage $expressionLanguage = null)
+    {
+        if (null === $expressionLanguage) {
```
Looks strange to me. I think I would make it a required constructor attribute because it cannot work without any `ExpressionLanguage` instance.
I would extract the code that creates a "secure" EL instance so it can be reused by someone else, if requried. Or make it a factory method at least. Wdyt?

m-vo replied:
:+1: for making it a required argument.

I'm not sure about the 'secure' bit though - does it ever make sense to have the `constant` function not disabled/overwritten? I have the feeling people would just forget to disable it when injecting their custom version. Or everyone would have to cater for it if there are multiple decorators in place. :thinking:

Toflar replied:
Yeah well...that's an argument that's basically true for everything. You could also just swap our authenticators for custom ones that do wrong checks and then the whole login setup is insecure :D 

You wanted to register your own functions, right? Then we need to register that EL as a service anyway and give you the option to tag your services. Afaik there's even something like an `ExpressionFunctionProviderInterface`.

m-vo replied:
Yes, we might want to inject a list of tagged providers instead and register them with `$expressionLanguage->registerProvider();`. Otherwise the only extension-friendly way of extending the EL is decorating the simple token parser.

Toflar commented on core-bundle/src/SimpleTokens/Parser.php:
```diff
@@ -0,0 +1,231 @@
+        );
+
+        $this->expressionLanguage = $expressionLanguage;
+        $this->logger = $logger;
```
Guess we could just make that service logger aware instead of having a constructor attribute?

m-vo replied:
eh, sure!

Toflar reviewed (changes requested):
Nice! Just a few conceptual remarks :)

m-vo committed:
use LoggerAware pattern

m-vo committed:
use LoggerAware pattern

m-vo committed:
rename service to contao.util.simple_token_parser

 - move class to Contao\CoreBundle\Util\SimpleTokenParser
 - move tests to Contao\CoreBundle\Tests\Util\SimpleTokenParserTest

m-vo committed:
rename service to contao.util.simple_token_parser

 - move class to Contao\CoreBundle\Util\SimpleTokenParser
 - move tests to Contao\CoreBundle\Tests\Util\SimpleTokenParserTest

m-vo committed:
use a custom expression language and allow registering extensions

Extensions need to implement the `ExpressionFunctionProviderInterface` and must be tagged with `contao.simple_token_extension`.

m-vo committed:
use a custom expression language and allow registering extensions

Extensions need to implement the `ExpressionFunctionProviderInterface` and must be tagged with `contao.simple_token_extension`.

m-vo committed:
add functional test for registering custom simple token extensions

m-vo committed:
add functional test for registering custom simple token extensions

m-vo commented:
@Toflar I implemented a way for extensions to add their functions (see updated description). Imo this way it's pretty easy to extend. Wdyt?

Toflar commented:
That's about what I had in mind! Nice :) Thinking about it, now that we have two classes, maybe we should move things into their own namespace anyway? But as a subnamespace of `Util`? What's your opinion on that?
Other than that, I think we're super close (check the CS issues or just run `composer cs-fixer -- --fix`)

m-vo commented:
> But as a subnamespace of Util? 

Thought about this as well. `Util\SimpleToken\Parser` but without a 'service namespace' (so still `contao.utill.simple_token_parser`)?

> check the CS issues

Do you have an idea on how to work around the `===` <->`==` conversions or how to compare with type juggling in any other manageable way?

Toflar commented:
Ah, that's this. No, I don't. I checked but apparently there's no way to skip certain lines? /cc @leofeyer

leofeyer commented:
I don't know yet, but I will check.

ausi commented on core-bundle/src/Util/SimpleTokenParser.php:
```diff
@@ -0,0 +1,213 @@
+        switch ($operator) {
+            case '==':
+                // todo: find a way to stop the CS fixer to change this to ===
+                return $tokenValue == $value;
```
```suggestion
                return \in_array($tokenValue, [$value]);
```
This would be not allowed too by the CS fixer i guess?

fritzmg replied:
The CS fixer will probably change it to `\in_array($tokenValue, [$value], true)`, which is the same as `$tokenValue === $value`.

ausi replied:
```suggestion
                switch ($tokenValue) {
                    case $value:
                        return true;
                    default:
                        return false;
                }
```

m-vo replied:
If I explicitly set the third parameter to false it works. :smile:

m-vo replied:
Nice find!

Toflar replied:
It can't be optimized by opcode anymore though :) So might as well kill the global function namespace but not sure if the fixer accepts that then :D

m-vo replied:
The `\` prefix gets added automatically.

m-vo committed:
prevent weak type comparison to be removed by the cs fixer

m-vo committed:
prevent weak type comparison to be removed by the cs fixer

m-vo committed:
CS

m-vo committed:
CS

m-vo commented:
What do you others think about the namespace?

Toflar commented on core-bundle/src/Util/SimpleTokenParser.php:
```diff
@@ -0,0 +1,213 @@
+
+        [, $token, $operator, $value] = $matches;
+
+        if (!\array_key_exists($token, $data)) {
```
These two statements can be combined into one `if` statement.

m-vo replied:
Not sure I understand? :monocle_face:

Toflar replied:
Ah, the diff was misleading. All good sorry.

m-vo replied:
Or do you mean something like this?
```php
if(null === ($tokenValue = $data[$token] ?? null)) {
    if (null !== $this->logger) {
        $this->logger->log(LogLevel::INFO, sprintf('Tried to evaluate unknown simple token "%s".', $token));
    }

    return false;
}
```

Edit: which wouldn't be correct because `$tokenValue` can be anything

Toflar reviewed (changes requested):
Nice work! Apart from my last nitpick and the namespace discussion I think this would be ready then.

Toflar reviewed (dismissed):
I'm okay with the namespace choice.

aschempp commented on core-bundle/src/Util/SimpleTokenParser.php:
```diff
@@ -0,0 +1,213 @@
+     */
+    private $expressionLanguage;
+
+    public function __construct(ExpressionLanguage $expressionLanguage)
```
Is there no interface to typehint against?

m-vo replied:
No, I don't think so.

aschempp commented on core-bundle/src/Resources/config/services.yml:
```diff
@@ -815,3 +815,15 @@ services:
             - '@contao.framework'
             - '@contao.routing.scope_matcher'
+
+    contao.util.simple_token_parser:
```
We decided to start using the class name for services. This will also enable autowiring for this class.

m-vo replied:
Did we really? I thought we would only use this for Controllers and Helpers where we want to use autowiring ourselves - but I may confuse something? 

I thought using the class name in a bundle is [bad practice](https://symfony.com/doc/current/bundles/best_practices.html#services). We could introduce an interface and use this instead (+ make the class final) but imo that's not worth the overhead?

aschempp commented on core-bundle/src/Util/SimpleTokenParser.php:
```diff
@@ -0,0 +1,213 @@
+    /**
+     * Parse simple tokens.
+     *
+     * @param string $subject The string to be parsed
```
There‘s no need for theses docblocks, as the argument name and type are already given by PHP7

m-vo replied:
I know, but what's with the ones adding descriptions? Would we drop the typehint only (`@param $subject The string to be parsed`)? And if so, could the CS fixer do this? :smile:

Toflar replied:
`The string to be parsed` is irrelevant anyway. It's obvious.

aschempp commented on core-bundle/src/Util/SimpleTokenParser.php:
```diff
@@ -0,0 +1,213 @@
+
+        if (!\array_key_exists($token, $data)) {
+            if (null !== $this->logger) {
+                $this->logger->log(LogLevel::INFO, sprintf('Tried to evaluate unknown simple token "%s".', $token));
```
Wasn’t this logged to the Contao backend previously?

m-vo replied:
It was taken from here:
https://github.com/contao/contao/blob/e7e0054dfc669f13e176217a1ed57282e2e7d987/core-bundle/src/Resources/contao/library/Contao/StringUtil.php#L644-L647

aschempp commented on core-bundle/src/Resources/config/services.yml:
```diff
@@ -815,3 +815,15 @@ services:
+        class: Contao\CoreBundle\Util\SimpleTokenParser
+        public: true
+        arguments:
+            - '@contao.util.simple_token_parser.expression_language'
```
Should we tag the logger to use the correct channel?

m-vo replied:
You mean this, right?
```
        tags:
            - { name: monolog.logger, channel: contao }
```

leofeyer replied:
The `BrokenLinkCheckerSubscriber` and the `SearchIndexSubscriber` also implement the `LoggerAwareInterface`. Should we tag them as well?

https://github.com/contao/contao/blob/5c6387596f18d6e2ae0fbf7ed39483a149d493c3/core-bundle/src/Crawl/Escargot/Subscriber/BrokenLinkCheckerSubscriber.php#L30

m-vo committed:
improve docblock and parameter name verbosity

m-vo committed:
improve docblock and parameter name verbosity

m-vo commented:
Failing tests are unrelated btw.

leofeyer committed:
Merge branch 'master' into feature/simple-token-parser-service

m-vo commented:
@aschempp What should be done here to unlock this PR?

leofeyer committed:
Merge branch 'master' into feature/simple-token-parser-service

leofeyer commented on core-bundle/tests/Functional/SimpleTokenExtensionTest.php:
```diff
@@ -0,0 +1,30 @@
+
+use Symfony\Bundle\FrameworkBundle\Test\KernelTestCase;
+
+class SimpleTokenExtensionTest extends KernelTestCase
```
This does not seem like a functional test that requires a running system. A regular unit test would suffice IMHO.

m-vo replied:
How would this work? 

see https://github.com/contao/contao/blob/46f5475e4d4707631c1f173c7429bc3dd1820f39/core-bundle/tests/Functional/app/config/services.yml

It tests adding a custom extension provider.

Toflar replied:
I like the additional functional test :)

leofeyer replied:
We do not have to test whether the Symfony container registers the tagged providers IMHO. Testing the DI container is Symfony's responsibility and not ours.

m-vo replied:
Imo the test documents nicely how this feature can/should be used (and ensures it's intended usage works). I like this in general (if reading the tests explains the code) but I can drop it as well if you prefer.

leofeyer replied:
@contao/developers WDYT?

leofeyer replied:
As discussed in Mumble on July 2nd, this should be a unit test. 🙈

m-vo replied:
Done

leofeyer committed:
Use the class names as service ID

leofeyer committed:
Use the class names as service ID

leofeyer commented on core-bundle/tests/Functional/Util/SimpleToken/StringExtensionProvider.php:
```diff
@@ -0,0 +1,26 @@
+    public function getFunctions(): array
+    {
+        return [
+            ExpressionFunction::fromPhp('strtoupper'),
```
What do we need this for?

m-vo replied:
It's a custom extension for the simple token expression language (this being a fixture for the related functional test).

Toflar replied:
This is just a functional test, that's correct :)

m-vo replied:
Should this go in the `Fixtures` dir?

leofeyer committed:
Run the CS fixer

leofeyer committed:
Run the CS fixer

leofeyer committed:
Merge branch 'master' into feature/simple-token-parser-service

Toflar commented:
No, they get a different logger!

m-vo committed:
migrate functional test to unit test + inline fixture class

m-vo committed:
migrate functional test to unit test + inline fixture class

leofeyer commented:
It seems that you are trying to use a file from the manager-bundle in the core-bundle:

```
1) Contao\CoreBundle\Tests\Util\SimpleTokenParserTest::testParseSimpleTokenWithCustomExtensionProvider
Error: Class 'Contao\ManagerBundle\Tests\Fixtures\IteratorAggregateStub' not found

/home/runner/work/contao/contao/core-bundle/tests/Util/SimpleTokenParserTest.php:536
/home/runner/work/contao/contao/core-bundle/vendor/phpunit/phpunit/phpunit:61
```

m-vo committed:
Merge branch 'feature/simple-token-parser-service' of github.com:m-vo/contao into feature/simple-token-parser-service

m-vo committed:
use IteratorAggregateStub from core-bundle instead of manager-bundle

leofeyer commented:
Thank you @m-vo.
